LOOP-1406: Add authentication when saving therapy settings

rickpasetto opened the pull request:
This refactors a bit of the "saving" logic to have common code for presenting the authentication challenge, and then presenting an alert if the challenge fails.

https://tidepool.atlassian.net/browse/LOOP-1406

Commit:
LOOP-1406: Add authentication when saving therapy settings

This refactors a bit of the "saving" logic to have common code for presenting the authentication challenge, and then presenting an alert if the challenge fails.

Commit:
Don't auth challenge in acceptance flow

mpangburn commented on LoopKitUI/Extensions/Environment+Authenticate.swift:
```diff
@@ -0,0 +1,41 @@
+import LocalAuthentication
+import SwiftUI
+
+public typealias Completion<Value> = (Result<Value, Error>) -> Void
```
[nit] I don't think this typealias is necessary

rickpasetto replied:
Ok.  I think someday we may want it ;)

mpangburn commented on LoopKitUI/Extensions/Environment+Authenticate.swift:
```diff
@@ -0,0 +1,41 @@
+public typealias AuthenticationChallenge = (_ description: String, _ completion: @escaping Completion<Void>) -> Void
+public extension Result where Success == Void {
+    static var success: Result {
+        return Result.success(Void())
```
[nit] Typically an instance of Void is spelled `()` (the empty tuple)

mpangburn commented on LoopKitUI/Extensions/Environment+Authenticate.swift:
```diff
@@ -0,0 +1,41 @@
+private struct AuthenticationChallengeKey: EnvironmentKey {
+    static let defaultValue: AuthenticationChallenge = { authenticationChallengeDescription, completion in
+        let context = LAContext()
+        if context.canEvaluatePolicy(.deviceOwnerAuthentication, error: nil) {
```
Should have an `else` branch that calls `completion` successfully if the policy can't be evaluated

rickpasetto replied:
ooh good catch!

mpangburn commented on LoopKitUI/Extensions/Environment+Authenticate.swift:
```diff
@@ -0,0 +1,41 @@
+                                   localizedReason: authenticationChallengeDescription,
+                                   reply: { (success, error) in
+                                    DispatchQueue.main.async {
+                                        completion(success ? .success : .failure(error ?? UnknownError()))
```
[nit] Perhaps worth asserting that `error` is non-nil if `success` is `false`

mpangburn commented on LoopKitUI/Extensions/Environment+Authenticate.swift:
```diff
@@ -0,0 +1,41 @@
+        return Result.success(Void())
+    }
+}
+public struct UnknownError: Swift.Error { }
```
I think this can be `fileprivate`

mpangburn commented on LoopKitUI/Views/ScheduleEditor.swift:
```diff
@@ -351,7 +349,22 @@ struct ScheduleEditor<Value: Equatable, ValueContent: View, ValuePicker: View, A
+        authenticate(authenticationChallengeDescription) {
+            switch $0 {
+            case .success: self.continueSaving()
+            case .failure(let error): self.presentedAlert = .saveError(error)
```
This suggestion might sound odd, but I'm actually not sure we should present an alert when authentication fails here. Since an incorrect biometry reading keeps the user in the authentication context; generally they proceed only by 1) correctly authenticating or 2) cancelling. We can defer communication of auth errors to the auth modal itself.

rickpasetto replied:
I don't know... I think this is a design call.  Normally I'd agree, but we're being ultra-conservative when it comes to communicating to the user what happened.  For example, if the user cancels the authentication, this will pop up an error:
![Simulator Screen Shot - iPhone SE (2nd generation) - 2020-08-03 at 15 07 30](https://user-images.githubusercontent.com/308993/89231877-165c0d00-d59b-11ea-8093-5e8c8e0dc2e9.png)

Pedantic?  Yes...but maybe not a bad idea?  I suggest we try leaving it as-is here and see how design review goes.  WDYT?

mpangburn commented on LoopKitUI/Views/Settings Editors/PresentedAlert.swift:
```diff
@@ -0,0 +1,46 @@
+
+import SwiftUI
+
+enum PresentedAlert {
```
If we defer auth errors to the modal, we can return this to the ScheduleEditor namespace. If you want to extract it, I think we should rename it for clarity; something like SettingSaveAlert.

rickpasetto replied:
I'll rename it... but I kind of like it as a reusable enum...

mpangburn reviewed (commented):
The general approach is nice! I believe we can simplify by deferring auth error presentation to the iOS modal.

Commit:
PR Feedback

Commit:
Merge branch 'dev' into rick/LOOP-1406-authenticate

Commit:
Merge branch 'dev' into rick/LOOP-1406-authenticate

Commit:
Merge branch 'rick/LOOP-1406-authenticate' of https://github.com/tidepool-org/LoopKit into rick/LOOP-1406-authenticate

Commit:
Oops.  PR feedback.

novalegra commented on LoopKitUI/Extensions/Environment+Authenticate.swift:
```diff
@@ -0,0 +1,45 @@
+                                    }
+            })
+        } else {
+            assert(error != nil)
+            completion(.failure(error ?? UnknownError()))
```
[nit/q] If we're asserting the error is `nil` here, then would it make sense to force unwrap it on the next line?

mpangburn replied:
I could see it either way; the assert won't be triggered in release builds, so if the framework had a bug where the assertion failed having a fallback isn't unreasonable.

rickpasetto replied:
`assert` is compiled out in production code (https://developer.apple.com/documentation/swift/1541112-assert)
(note: `precondition` is not)
force-unwrapping would cause a crash in production code.  I felt it was unnecessary to do that, though it would not be unreasonable given that it would indicate Apple's APIs would violate their own contract, which is unlikely.  I could go either way.

novalegra reviewed (approved):
One q, otherwise LGTM!

mpangburn commented on LoopKitUI/Extensions/Environment+Authenticate.swift:
```diff
@@ -0,0 +1,45 @@
+import SwiftUI
+
+public typealias AuthenticationChallenge = (_ description: String, _ completion: @escaping (Result<Void, Error>) -> Void) -> Void
+public extension Result where Success == Void {
```
Ah, just noticed this extension is public. I'm not sure it makes sense for LoopKitUI to vend this as public API.

mpangburn commented on LoopKitUI/Extensions/Environment+Authenticate.swift:
```diff
@@ -0,0 +1,45 @@
+                                    }
+            })
+        } else {
+            assert(error != nil)
+            completion(.failure(error ?? UnknownError()))
```
I'm not sure failure is the correct outcome if the policy can't be evaluated. What if the error is `biometryNotAvailable` or `biometryNotEnrolled`? If a user doesn't have auth set up, they should still be able to save their settings.

The precedent here (in bolusing) is to proceed as successful if the policy can't be evaluated.

rickpasetto replied:
Good point.  I'll ask Tejasvi in Slack what the behavior should be.

mpangburn commented on LoopKitUI/Views/ScheduleEditor.swift:
```diff
@@ -351,7 +349,22 @@ struct ScheduleEditor<Value: Equatable, ValueContent: View, ValuePicker: View, A
+        authenticate(authenticationChallengeDescription) {
+            switch $0 {
+            case .success: self.continueSaving()
+            case .failure(let error): self.settingSaveAlert = .saveError(error)
```
I'll continue to push back on presenting an error here. All error states are communicated by the auth modal. If the user pushes cancel, they should not see a modal that tells them they just cancelled. I weighed in on the Slack thread too; seems that Pete agrees we don't need to handle the presentation of an error here.

rickpasetto replied:
Okay, I will remove it.  That said, would you be okay with my leaving this new `SettingSaveAlert` in place as a generally-usable facility?

mpangburn replied:
Yeah, I don't see a problem with that

rickpasetto replied:
;) I got rid of it.  Sorry...Please take another look, @mpangburn ?

mpangburn reviewed (commented):
If auth is unavailable, I believe the outcome should be `success`.

I continue to think we don't need to handle presentation of auth errors, but won't block on that.

Commit:
Remove SettingSaveAlert

Now that we don't want to show an alert due to auth failure, it is unnecessary.

mpangburn commented on LoopKitUI/Views/Settings Editors/DeliveryLimitsEditor.swift:
```diff
@@ -277,7 +276,7 @@ public struct DeliveryLimitsEditor: View {
             switch $0 {
             case .success: self.continueSaving()
-            case .failure(let error): self.settingSaveAlert = .saveError(error)
+            case .failure(let error): print("Authentication failed: \(error)")
```
`print` can be deleted

mpangburn commented on LoopKitUI/Views/Settings Editors/SuspendThresholdEditor.swift:
```diff
@@ -190,7 +190,7 @@ public struct SuspendThresholdEditor: View {
             switch $0 {
             case .success: self.continueSaving()
-            case .failure(let error): self.settingSaveAlert = .saveError(error)
+            case .failure(let error): print("Authentication failed: \(error)")
```
`print` can be deleted

mpangburn commented on LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift:
```diff
@@ -270,7 +264,7 @@ public struct InsulinModelSelection: View, HorizontalSizeClassOverride {
             switch $0 {
             case .success: self.continueSaving()
-            case .failure(let error): self.settingSaveAlert = .saveError(error)
+            case .failure(let error): print("Authentication failed: \(error)")
```
`print` can be deleted

mpangburn commented on LoopKitUI/Views/Settings Editors/CorrectionRangeOverridesEditor.swift:
```diff
@@ -237,7 +238,7 @@ public struct CorrectionRangeOverridesEditor: View {
             switch $0 {
             case .success: self.continueSaving()
-            case .failure(let error): self.settingSaveAlert = .saveError(error)
+            case .failure(let error): print("Authentication failed: \(error)")
```
`print` can be deleted

mpangburn reviewed (approved):
You can also remove the 'cancel' and 'ok' properties that were added to AlertContent now, I believe

Commit:
Get rid of extraneous print statements

Commit:
remove print statements

Commit:
Adds "cancel" button to all of the therapy settings editors

Per the slack thread, it was determined that implying that "back" would cancel the user's edited setting, using "Cancel" was determined to be more clear.  So I've added a cancel button to each editor, but it meant also wiring a cancel() function in so that the parent NavigationLink could navigate back.

rickpasetto commented:
Sorry to do this to you @mpangburn and @novalegra , but I received feedback on a slack thread that we need to change the "back" button behavior of the editors to show `Cancel` instead, so that it is clear users that their settings change will be discarded.  So, if you could please look at the new changes above, I'd really appreciate it. :)

Commit:
Another tweak to the "cancel" behavior

Another "implicit" design requirement was that the "Cancel" button only show if the value is edited, but that the "back" button should still show if it isn't.  This changes that in all screens.

Commit:
Merge branch 'dev' into rick/LOOP-1406-authenticate

rickpasetto commented:
@mpangburn @novalegra ...and, another update.  Double apologies again, but I now have learned that the expected behavior is for the "Cancel" button to only appear when the value has changed, otherwise it _should_ show the Back (previous screen) button.  So, I've updated it...please have another look.

novalegra commented on LoopKitUI/Extensions/Environment+Authenticate.swift:
```diff
@@ -0,0 +1,40 @@
+private struct AuthenticationChallengeKey: EnvironmentKey {
+    static let defaultValue: AuthenticationChallenge = { authenticationChallengeDescription, completion in
+        let context = LAContext()
+        var error:NSError?
```
[nit] Missing a space

novalegra commented on LoopKitUI/Views/ScheduleEditor.swift:
```diff
@@ -143,16 +149,31 @@ struct ScheduleEditor<Value: Equatable, ValueContent: View, ValuePicker: View, A
             return AnyView(wrappedPage)
         case .acceptanceFlow, .settings:
-            return AnyView(configurationPage)
+            return AnyView(contentWithCancel)
```
I don't think we want the cancel button to be in the acceptance flow - users should be able to go back and forth between screens (I discussed the "close" button issue with Melissa and Brooke, and we all agreed that the editors are too crowded to have one)

rickpasetto replied:
Good catch.  I'll fix it! 👍

novalegra commented on LoopKitUI/Views/Settings Editors/InsulinModelSelection.swift:
```diff
@@ -77,27 +103,44 @@ public struct InsulinModelSelection: View, HorizontalSizeClassOverride {
-            .navigationBarItems(leading: dismissButton)
+            content
+                .navigationBarItems(leading: dismissButton)
+                .navigationBarTitle(Text(TherapySetting.insulinModel.title), displayMode: .large)
```
There's no title for the other configurations of this screen, which makes it so no title appears in the acceptance flow or the new therapy settings screen.

rickpasetto replied:
I was thrown by the fact that sometimes the title "Insulin Model" got "stuck" sometimes when navigating back.  I thought this would fix it, but turns out this is wrong, too.  I can't reproduce it, so I'll put this back to where it used to be.  But I'll keep an eye out for it.

novalegra reviewed (changes requested):
There are some changes that affect titles/the acceptance flow that we should probably fix before this lands.

mpangburn reviewed (commented):
I think the presence of `cancel` being plumbed all the way through is a sign that it should go through the environment. You can pass it through using the `dismiss` key.

Commit:
PR Feedback

rickpasetto commented:
@mpangburn good idea.  I'll see if I can do that.

Commit:
Move the cancel closure to an environment

Thanks for the suggesiton, @mpangburn!  A lot cleaner...

Commit:
make the whole card tappable

novalegra commented on LoopKitUI/Views/ScheduleEditor.swift:
```diff
@@ -138,26 +141,43 @@ struct ScheduleEditor<Value: Equatable, ValueContent: View, ValuePicker: View, A
-                leading: cancelButton, // add in cancel button if modal
-                trailing: trailingNavigationItems
+            page
+                .navigationBarItems(
+                    leading: cancelButton, // add in cancel button if modal
+                    trailing: trailingNavigationItems
```
[nit] spacing

rickpasetto replied:
This is indented how Xcode wants it.  I could reduce the vertical space used...

novalegra replied:
Gotcha - no need to reduce vertical space :)

novalegra reviewed (approved):
LGTM, thanks Rick!

mpangburn reviewed (approved):
Looks great!
